course_contents/2_intro_to_python/notes/02. Loops.md

```diff
@@ -0,0 +1,159 @@
+    3. indented block of code that will run until the loop breaks
+
+An example of the properly written `while` loop:
+
+```py
+counter = 10
+while counter > 0:
+  print(f"{counter} seconds left until New Year!")
+  counter -= 1
+print("Happy New Year!")
+```
+
+An example of the `while` loop that will run until the program is manually stopped:
+
+```py
+counter = 10
+while counter > 0: # the counter never gets changed, causing the conditional to be always True
+  print(f"{counter} seconds left until New Year!")
+```
+
+### The `for` loop
+- for loop runs a given number of times
+- general syntax for `for` loop is:
+
+```py
+for variable in iterable:
+    block_of_code
+```
+- an iterable is an object that can be iterated over e.g. a list
+- an object is iterable if it can return the iterator object (more on that in the OOP section)
+
+An example of the `for` loop:
+
+```py
+numbers = [1,2,3,4,5]
+for i in numbers:
+    print(i)
+```
+
+**Extra resources**
+[How does the for loop work in Python?](https://python.tecladocode.com/3_continuous_countries/1_loops.html#how-does-the-for-loop-work-in-python)
+
+#### The `range()` method
+- the `range()` method returns a sequence of integers in a given range
+- arguments that are passed to the method are:
+    1. `start` - an integer marking the start of a sequence
+    2. `stop` - an integer marking the end of a sequence. Important - `range()` returns integers up to `stop-1`
+    3. `step` - an integer representing the size of the step i.e. the difference between consecutive integers in a sequence
+- the numbers are generated by starting from `start` and incrementing it by `step` until `stop-1` is reached
+- the only argument that has to be stated explicitly is the `stop`
+- if not specified, Python will implicitly set `start` to 0 and `step` to 1
+
+An example of the `range()` method used in a `for` loop:
+
+```py
+for i in range(5): #the range() method can be used to specify number of iterations of a loop
+  print(i) #the numbers printed will be: 0,1,2,3,4
+```
+
+### The `break` statement
+- the `break` statement is used to stop the loop from running when an external condition is reached
+- after the loop has been exited - the next line of code after it will be executed
+
+An example of using the `break` statement in the `while` loop:
+
+```py
+counter = 0
+while counter < 10:
+  print(counter)
+  if counter == 5:
+    break #the loop stops running when the counter is equal to 5
+  counter += 1
+print(f"That's all the numbers from 0 to {counter}!") #after the break, this line is executed
+```
+
+### The `continue` statement
+- the `continue` statement is used to skip a step in a loop and run the next iteration
+- no code after the `continue` statement will be run
+
+An example of the `continue` statement being used in a `for` loop:
+
+```py
+for i in range(5):
+    if i % 2 == 0: 
+        continue #if i is even, it won't be printed out
+    print(i)
```

Comment:
Same here, could we use the course example?

```py
cars = ["ok", "ok", "ok", "faulty", "ok", "ok"]

for status in cars:
    if status == "faulty":
        print("Found faulty car, skipping...")
        continue

    print(f"This car is {status}.")
    print("Shipping new car to customer!")
```

Reply:
^ In here I would be tempted to use something like this (a different example from the one `break` statement). Just to give a wider range of examples, and also this one is a bit more interesting.

Open to alternatives though, if you don't want to use the one from the course!

Reply:
Sure, I can change the example to this one 👍